sys/kern/src/startup.rs

```diff
@@ -118,7 +118,6 @@ pub unsafe fn start_kernel(tick_divisor: u32) -> ! {
     let first_task_index =
         crate::task::select(task_table.len() - 1, task_table);
 
-    crate::arch::apply_memory_protection(&task_table[first_task_index]);
     TASK_TABLE_IN_USE.store(false, Ordering::Release);
     crate::arch::start_first_task(
         tick_divisor,
```

Comment:
I would have like to put this one line below...

It's interesting that start_first_task was already being provided a mutable reference, but the compiler didn't care and dropped it. I'm more curious about the history here since it seems a mutable reference was necessary at some point.

Is the first task the idle task? Or is there a notion of this thread of context continuing past this line?

Reply:
You won't believe it, but as I was refactoring this, I got curious about the same thing and started playing with the code to see why the mutable reference is required. It looks like I can get rid of the mutable reference through the entire chain. I am still not sure why this mutable reference exists in the first place though. I will see if I can dig up some history.

> Is the first task the idle task? Or is there a notion of this thread of context continuing past this line?

First task could be anything that is marked with `START_AT_BOOT` flag in the app configuration: https://hubris.oxide.computer/reference/#_starting_the_first_tasks

This and `start_first_task` functions never return (return is defined as `!`).

Reply:
I'm not sure about the mutable reference either. I was going to clean it up as a part of a patch making the kernel clippy clean.

Reply:
> This and `start_first_task` functions never return (return is defined as `!`).

Thanks for the pointer. Good to know. Makes more sense then.

Reply:
[This](https://github.com/oxidecomputer/hubris/pull/582/commits/9cb4d9753c7a4c9f18fd993968dd49931a952c9e#diff-071eb4f76fe8d7a0d61bcc0454c6a2d03527130e5227e17f58dd758419469fef) seems to be the commit that introduced it. Still trying to make sense of it.

Reply:
Okay, so it looks like this is required for the [atomic store](https://doc.rust-lang.org/core/sync/atomic/struct.AtomicPtr.html#method.store)

This store method is called [here](https://github.com/oxidecomputer/hubris/blob/a6b0055273c067af60c0c76abde5a54c096f362e/sys/kern/src/arch/arm_m.rs#L655)

So, yeah I cannot get rid of the mutable reference.

Reply:
I wasn't suggesting to get rid of it. I was noting it was already there prior. I guess we just didn't need it in the riscv port?

Reply:
Looks like our [RV port hasn't yet switched to using atomic ops](https://github.com/rivosinc/hubris/blob/rivos/main/sys/kern/src/arch/riscv32.rs#L706).